File: test/amd64/scripts/run_pipes_test.py
#
# This file is part of libdebug Python library (https://github.com/libdebug/libdebug).
# Copyright (c) 2024 Roberto Alessandro Bertolini. All rights reserved.
# Licensed under the MIT license. See LICENSE file in the project root for details.
#

import unittest

from pwn import process


class RunPipesTest(unittest.TestCase):
    def test_binary_proxy(self):
        unpatched = process("binaries/run_pipes_test")

        unpatched.recvuntil(b"1.")

        # Option 1 should print the flag
        unpatched.sendline(b"1")
        self.assertIn(b"flag{provola}", unpatched.recvuntil(b"1."))

        # Option 2 should print the flag after the admin mode check
        unpatched.sendline(b"2")
        unpatched.sendline(b"admin")
        self.assertIn(b"flag{provola}", unpatched.recvuntil(b"1."))

        # Option 3 should print the flag after the signal handler
        unpatched.sendline(b"3")
        self.assertIn(b"flag{provola}", unpatched.recvuntil(b"1."))

        # Exit
        unpatched.sendline(b"4")

        unpatched.kill()

        patched = process(["python3", "scripts/run_pipes_test_script.py"])

        patched.recvuntil(b"1.")

        # Option 1 should print not the flag
        patched.sendline(b"1")
        self.assertIn(b"flag{nahmate}", patched.recvuntil(b"1."))

        # Option 2 should not print the flag after the admin mode check
        patched.sendline(b"2")
        patched.sendline(b"admin")
        self.assertIn(b"Wrong password", patched.recvuntil(b"1."))

        # Option 3 should not print the flag after the signal handler
        patched.sendline(b"3")
        self.assertIn(b"SIGPROVOLA", patched.recvuntil(b"1."))

        # Exit
        patched.sendline(b"4")

        patched.kill()

File: test/common/srcs/run_pipes_test.c
//
// This file is part of libdebug Python library (https://github.com/libdebug/libdebug).
// Copyright (c) 2024 Roberto Alessandro Bertolini. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for details.
//

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <signal.h>

#define SIGPROVOLA 25

const char admin_pwd[] = "admin";
const char flag[] = "flag{provola}";

void option_1()
{
    printf("The flag is: %s\n", flag);
}

void option_2()
{
    char input[128];

    fgets(input, sizeof(input), stdin);
    input[strcspn(input, "\n")] = 0;

    if (strcmp(input, admin_pwd) == 0)
    {
        printf("Welcome admin!\n");
        printf("The flag is: %s\n", flag);
    }
    else
    {
        printf("Wrong password!\n");
    }
}

void sigprovola_handler(int sig)
{
    printf("Wowsers! This fine piece of code developed by Io_no should have never reached this state!\n");
    printf("Stacktrace[0]: %p\n", __builtin_return_address(0));
    printf("Stacktrace[1]: %s\n", flag);
    printf("Stacktrace[2]: %p\n", __builtin_return_address(2));
}

int main()
{
    setvbuf(stdout, NULL, _IONBF, 0);

    // Register the signal handler
    signal(SIGPROVOLA, sigprovola_handler);

    int choice;

    while (1)
    {
        printf("Welcome to Io_no's personal flag management system!\n");
        printf("Choose an option:\n");
        printf("1. Print the flag\n");
        printf("2. Become admin and print the flag\n");
        printf("3. Raise a signal\n");
        printf("4. Exit\n");

        scanf("%d", &choice);
        fgetc(stdin);

        switch (choice)
        {
        case 1:
            option_1();
            break;
        case 2:
            option_2();
            break;
        case 3:
            raise(SIGPROVOLA);
            break;
        case 4:
            return 0;
        default:
            printf("Invalid choice!\n");
            break;
        }
    }
}